### PR TITLE
Add methods for adding, updating, retrieving and deleting properties

### PR DESCRIPTION
**Pull request by robocopkaka**
#### What does this PR do?
Add Property resource
#### Description of Task to be completed?
 - add method for adding new properties
  - add middleware for validating input for properties
  - add middleware for authenticating landlords
  - add route for adding properties
  - add tests
  - add methods for updating a property
  - add methods and tests for deleting a property
  - add methods and tests for retrieving a property
  - add methods and tests for retrieving properties

misc:
  - rename User to Landlord
  - rename Landlord to LandlordController in userController.ts
  - return all errors when validating schemas
#### How should this be manually tested?
A) To create a property, send:
```
curl -L -X POST 'http://localhost:8000/properties' \
-H 'token: actual token' \
-H 'Content-Type: application/json' \
--data-raw '{
	"name": "Test Property",
	"description": "Echi",
	"propertyType": "commercial",
	"address": "here",
	"category": "rent",
	"available": "false"
}'
```

Should return:
```.json
{
    "message": "Property added successfully",
    "property": {
        "id": 17,
        "name": "Test Property",
        "description": "Echi",
        "propertyType": "commercial",
        "address": "here",
        "features": [],
        "category": "rent",
        "available": false,
        "landlordId": 1,
        "uid": "ba0fd456-86d3-40f8-9cd6-63dea745c33b"
    }
}
```

B) To update a property, send

```
curl -L -X PATCH 'http://localhost:8000/properties/11' \
-H 'token: actualToken' \
-H 'Content-Type: application/json' \
--data-raw '{
	"name": "Edited name twice",
	"address": "edited here"
}'
```

Which should return:

```.json
{
    "message": "Property updated successfully",
    "property": {
        "id": 11,
        "name": "Edited name twice",
        "propertyType": "commercial",
        "description": "Echi",
        "address": "edited here",
        "features": [],
        "uid": "66f83de6-607f-4b98-97a7-b8baecbcafcb",
        "category": "rent",
        "available": false,
        "landlordId": 1
    }
}
```

C) Retrieve all properties
Send: 

```
curl -L -X GET 'http://localhost:8000/properties'
```

Which should return:

```.json
{
    "properties": [
        {
            "id": 2,
            "name": "Test Property",
            "propertyType": "commercial",
            "description": "Echi",
            "address": "here",
            "features": null,
            "uid": "d2e029c4-3ce8-43fa-aaa4-b0df4f274445",
            "category": "rent",
            "available": false,
            "landlordId": 1,
            "createdAt": "2020-04-12T04:52:58.593Z",
            "updatedAt": "2020-04-12T04:52:58.593Z"
        },
        {
            "id": 3,
            "name": "Test Property",
            "propertyType": "commercial",
            "description": "Echi",
            "address": "here",
            "features": null,
            "uid": "2bfcaff3-9035-4596-a30c-8672d916d12f",
            "category": "rent",
            "available": false,
            "landlordId": 1,
            "createdAt": "2020-04-12T05:27:41.558Z",
            "updatedAt": "2020-04-12T05:27:41.558Z"
        }
    ],
    "message": "Properties returned successfully"
}
```

D) Return a single property
Send

```
curl -L -X GET 'http://localhost:8000/properties/2'
```

Which should return:

```.json
{
    "property": {
        "id": 2,
        "name": "Test Property",
        "propertyType": "commercial",
        "description": "Echi",
        "address": "here",
        "features": [],
        "uid": "d2e029c4-3ce8-43fa-aaa4-b0df4f274445",
        "category": "rent",
        "available": false,
        "landlordId": 1
    },
    "message": "Property returned successfully"
}
```

E) To delete a property,
Send:

```
curl -L -X DELETE 'http://localhost:8000/properties/3' \
-H 'token: actualTokens'
```

Which should return:

```.json
{
    "message": "Property deleted successfully"
}
```

#### Any background context you want to provide?
If you experience any issues running tests, connect to your test database through your terminal and try the most accepted answer here - https://stackoverflow.com/questions/3327312/how-can-i-drop-all-the-tables-in-a-postgresql-database

#### What are the relevant pivotal tracker stories? (if applicable)
https://www.pivotaltracker.com/story/show/171429253
https://www.pivotaltracker.com/story/show/172163127
https://www.pivotaltracker.com/story/show/171429387

### CODE REVIEW

File: src/controllers/propertyController.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        address,
+        features,
+        category,
+        available,
+        landlordId
+      });
+      returnedObject.message = 'Property added successfully';
+      returnedObject.property = property.toJson()
+
+      res.status(201).json(returnedObject)
```
Comment (fire-cracker) [inline src/controllers/propertyController.ts]: Why not just do  res.status(201).json({message: 'Property added successfully', property})

File: src/controllers/propertyController.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    } catch (error) {
+      returnedObject.message = 'Internal server error'
+      res.status(500).json(returnedObject)
```
Comment (fire-cracker) [inline src/controllers/propertyController.ts]: res.status(500).json({message: 'Internal server error'})

File: src/controllers/propertyController.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      }
+
+      const propertyFound = await Property.findOne({
+        where: { id },
+      });
+
+      if (!propertyFound) {
+        returnedObject.message = 'Property not found'
+        return res.status(404).json(returnedObject)
+      }
```
Comment (fire-cracker) [inline src/controllers/propertyController.ts:52]: Same here. I feel mutating the object is redundant

File: src/controllers/propertyController.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+      const id = parseInt(req.params.id, 10);
+      const updateData : any = {};
+
+      if (name) { updateData.name = name }
+      if (description) { updateData.description = description }
+      if (propertyType) { updateData.propertyType = propertyType }
+      if (address) { updateData.address = address }
+      if (features) { updateData.features = features }
+      if (category) { updateData.category = category }
+      if (available !== null) { updateData.available = available }
```
Comment (fire-cracker) [inline src/controllers/propertyController.ts]: Check for the validity of the request body is a great idea but the middleware should check for this, not the controller.

File: src/controllers/propertyController.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+      updateData.returning = true;
+      updateData.plain = true;
+
+      if (Object.entries(req.body).length === 0) {
+        returnedObject.message = 'Please enter a value to be updated'
+        return res.status(400).json(returnedObject);
+      }
```
Comment (fire-cracker) [inline src/controllers/propertyController.ts:35]: Middleware.

File: src/controllers/propertyController.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        where: { id },
+      });
+
+      if (!propertyFound) {
+        returnedObject.message = 'Property not found'
+        return res.status(404).json(returnedObject)
+      }
+
+      if (parseInt(req.headers.landlordId, 10) !== propertyFound.landlordId) {
+        returnedObject.message = 'You don\'t own this property';
```
Comment (fire-cracker) [inline src/controllers/propertyController.ts]: How about "You are unauthorized to update this property"

File: src/controllers/propertyController.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+      returnedObject.message = 'Internal server error';
+      return res.status(500).json(returnedObject);
+    }
+  }
+
+  public static async delete (req, res) {
+    const returnedObject : any = {};
+    try {
+      const { landlordId } = req.headers;
+      const { id } = req.params;
```
Comment (fire-cracker) [inline src/controllers/propertyController.ts]: const { headers: {landlordId},  params: {id} } = req;

File: src/middleware/validation/validationHandler.ts [hunk 8]
```diff
@@ -1,10 +1,13 @@
 export const validationHandler = async (req, schema, res, next) => {
-  const { error } = await schema.validate(req);
+  const { error } = await schema.validate(req, { abortEarly: false });
   if (error) {
+    const messages = [];
+    error.details.map((err) => {
```
Comment (fire-cracker) [inline src/middleware/validation/validationHandler.ts]: This looks great but I think it is expensive

File: src/middleware/validation/validationHandler.ts [hunk 9]
```diff
@@ -1,10 +1,12 @@
 export const validationHandler = async (req, schema, res, next) => {
-  const { error } = await schema.validate(req);
+  const { error } = await schema.validate(req, { abortEarly: false });
   if (error) {
+    const messages = error.details.map((err) => {
```
Comment (robocopkaka) [inline src/middleware/validation/validationHandler.ts:4]: @fire-cracker Modified this and removed the call to `push`

File: src/middleware/validatePropertyRequestBody.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  if (features) { updateData.features = features }
+  if (category) { updateData.category = category }
+  if (available !== null) { updateData.available = available }
+
+  if (Object.entries(req.body).length === 0) {
+    return res.status(400).json({ message: 'Please enter a value to be updated' });
+  }
+
+  req.body['updateData'] = updateData;
+
```
Comment (fire-cracker) [inline src/middleware/validatePropertyRequestBody.ts]: Great job, though I don't think this is necessary. Checking for the length of the body in the controller is enough. If the body Is empty then an error is thrown.

**Review by fire-cracker — CHANGES_REQUESTED** [review]

**Review by fire-cracker — APPROVED** [review]

**Pull request merged**